erts/emulator/beam/jit/instr_common.cpp

```diff
@@ -617,18 +617,50 @@ void BeamModuleAssembler::emit_i_make_fun(const ArgVal &Fun,
     a.mov(getXRef(0), RET);
 }
 
-void BeamModuleAssembler::emit_get_list(const ArgVal &Src,
+void BeamModuleAssembler::emit_get_list(const x86::Gp src,
                                         const ArgVal &Hd,
                                         const ArgVal &Tl) {
-    mov_arg(ARG1, Src);
+    x86::Gp boxed_ptr = emit_ptr_val(src, src);
 
-    x86::Gp boxed_ptr = emit_ptr_val(ARG1, ARG1);
+    switch (ArgVal::register_relation(Hd, Tl)) {
+    case ArgVal::Relation::consecutive: {
+        comment("(moving head and tail together)");
+        x86::Mem dst_ptr = getArgRef(Hd);
+        x86::Mem src_ptr = getCARRef(boxed_ptr);
+        dst_ptr.setSize(16);
+        src_ptr.setSize(16);
+        a.movups(x86::xmm0, src_ptr);
+        a.movups(dst_ptr, x86::xmm0);
+        break;
+    }
+    case ArgVal::Relation::reverse_consecutive: {
+        if (! hasCpuFeature(x86::Features::kAVX)) {
+            goto fallback;
+        }
 
-    a.mov(ARG2, getCARRef(boxed_ptr));
-    a.mov(ARG3, getCDRRef(boxed_ptr));
+        comment("(moving and swapping head and tail together)");
+        x86::Mem dst_ptr = getArgRef(Tl);
+        x86::Mem src_ptr = getCARRef(boxed_ptr);
+        dst_ptr.setSize(16);
+        src_ptr.setSize(16);
+        a.vpermilps(x86::xmm0, src_ptr, 0x4E); /* Load and swap */
```

Comment:
I think `vpermilpd` with a control byte of 1 is a bit easier to follow in this case. It's otherwise identical with regards to size, speed, latency, etc so it's just a matter of taste.

```
IF (imm8[0] = 0) THEN DEST[63:0]←SRC1[63:0]
IF (imm8[0] = 1) THEN DEST[63:0]←SRC1[127:64]
IF (imm8[1] = 0) THEN DEST[127:64]←SRC1[63:0]
IF (imm8[1] = 1) THEN DEST[127:64]←SRC1[127:64]
DEST[MAXVL-1:128]←0
```